### ISSUE TITLE
Enabling runtime metrics results in assertion error with python 3.9 with gunicorn and gevent

### ISSUE THREAD

**Issue by lawrenceong** [DataDog/dd-trace-py#2610]
### Which version of dd-trace-py are you using?

0.50.0rc6

### Which version of pip are you using?

Pip 21.1.3 with Python 3.9.6

### Which version of the libraries are you using?

Minimum `requirements.txt` file:

```
ddtrace==0.50.0rc6
flask==2.0.1
gunicorn==20.1.0
gevent==21.1.2
```

### How can we reproduce your problem?

Create a very simple flask app.  For example (`app.py`):

```
from flask import Flask
app = Flask(__name__)

@app.route('/')
def hello_world():
    return 'Hello, World!'
```

Install the libraries by running `pip install -r requirements.txt`, then run:

```
DD_GEVENT_PATCH_ALL=true; export DD_GEVENT_PATCH_ALL
DD_RUNTIME_METRICS_ENABLED=true; export DD_RUNTIME_METRICS_ENABLED

ddtrace-run gunicorn --bind 0.0.0.0:8082 -k gevent -w 1 --threads 1 app:app
```

### What is the result that you get?

```
[2021-07-07 19:17:49 +1000] [77422] [INFO] Starting gunicorn 20.1.0
[2021-07-07 19:17:49 +1000] [77422] [INFO] Listening at: http://0.0.0.0:8082 (77422)
[2021-07-07 19:17:49 +1000] [77422] [INFO] Using worker: gevent
Traceback (most recent call last):
  File "src/gevent/_abstract_linkable.py", line 287, in gevent._gevent_c_abstract_linkable.AbstractLinkable._notify_links
  File "src/gevent/_abstract_linkable.py", line 333, in gevent._gevent_c_abstract_linkable.AbstractLinkable._notify_links
AssertionError: (None, <callback at 0x105c1e6c0 args=([],)>)
2021-07-07T09:17:49Z <callback at 0x105c1e6c0 args=([],)> failed with AssertionError
```

### What is the result that you expected?

AssertionError should not be happening.

### Other notes

* Setting `DD_RUNTIME_METRICS_ENABLED=false` will stop the AssertionError from happening
* Leaving `DD_RUNTIME_METRICS_ENABLED=true` and using Python 3.8 will also stop the AssertionError from happening

**Comment by majorgreys** [comment]
@lawrenceong thank you for sharing this reproduction. I have created a gist for the code. We will investigate the root cause and update here with any progress.

https://gist.github.com/majorgreys/112dafbebe05c0f1be86878166b7476c

**Comment by majorgreys** [comment]
@lawrenceong Can you try the following as a workaround?

Add the following to a gunicorn config file (e.g. `gunicorn.conf.py`):

```python
def post_worker_init(worker):
    import ddtrace.bootstrap.sitecustomize
    import ddtrace
    from ddtrace.runtime import RuntimeMetrics
    RuntimeMetrics.disable()
    RuntimeMetrics.enable(tracer=ddtrace.tracer, dogstatsd_url=ddtrace.tracer._dogstatsd_url, flush_interval=10)
```

You will need to remove `ddtrace-run` from your command. Instead you would need to specify the config file:

```
gunicorn -c gunicorn.conf.py --bind 0.0.0.0:8082 -k gevent -w 1 --threads 1 app:app
```

**Comment by lawrenceong** [comment]
Thanks @majorgreys, workaround looks good.  Do we need `RuntimeMetrics.disable()` before the enable?

**Comment by majorgreys** [comment]
@lawrenceong I'm glad the workaround did the trick. Yes, for now you will need it otherwise the subsequent enable will not have the intended result. I've started a fix in #2613 which will address this issue. We'll also update the gunicorn documentation accordingly.

**Comment by majorgreys** [comment]
The #2613 fix was released in a patch for both 0.49 and 0.50. Also the documentation for gunicorn has been updated in #2628.
